Update old type annotation syntax to new TypeHints

nicklaustrup opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
Fixes #27845


#### Brief description of what is fixed or changed
 This PR modernizes type annotations in the SymPy codebase by:

- Convert "# type: X" comments to "variable: X" syntax.
- Replace typing aliases (tDict, tTuple, etc.) with built-in types.
- Add `from __future__ import annotations` for Python 3.8 compatibility.
- Remove redundant None initializations used only for typing.
- Convert Optional[X] to X | None and Union[A, B] to A | B.
- Added ' | None' to type for 'is_real: bool' in stats/rv.py::PSpace
- Updated 'Expr' to Expr in core/evalf.py

Only remaining typing tType:
- from typing import Sequence as tSequence
- In sympy/stats/stochastic_process_types.py

#### Other comments
Files modified:
```
$ git diff --name-only origin/master -- "*.py"
 sympy/__init__.py
sympy/abc.py
sympy/codegen/ast.py
sympy/codegen/fnodes.py
sympy/concrete/products.py
sympy/concrete/summations.py
sympy/core/evalf.py
sympy/core/exprtools.py
sympy/external/gmpy.py
sympy/external/pythonmpq.py
sympy/functions/combinatorial/numbers.py
sympy/functions/elementary/complexes.py
sympy/functions/elementary/exponential.py
sympy/functions/elementary/integers.py
sympy/functions/elementary/trigonometric.py
sympy/matrices/common.py
sympy/matrices/dense.py
sympy/matrices/matrixbase.py
sympy/parsing/sympy_parser.py
sympy/physics/units/definitions/dimension_definitions.py
sympy/physics/units/unitsystem.py
sympy/polys/domains/gaussiandomains.py
sympy/polys/matrices/domainmatrix.py
sympy/printing/printer.py
sympy/solvers/diophantine/diophantine.py
sympy/solvers/ode/single.py
sympy/stats/rv.py
sympy/stats/stochastic_process_types.py
sympy/tensor/array/expressions/array_expressions.py
sympy/tensor/array/expressions/from_array_to_matrix.py
sympy/tensor/tensor.py
```

#### Release Notes
<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

nicklaustrup committed:
Removed '# type:' comments from sympy/matrices/common

Updated sympy/matrices/common.py to use modern
PEP 526 styling by adding type hinting. Also
added an import for __future__ annotations in
so that code is compatible with python 3.8.

nicklaustrup committed:
Remove old type annotation syntax (Issue #27845)

This PR modernizes type annotations in /matrices/matrixbase.py module.
Changes include:
- Converting type comments to PEP 526 variable annotations
- Adding `from __future__ import annotations` for Python 3.8 compatibility
- Removing redundant None initializations used only for type hinting

Tests pass, and code quality checks have been performed.

nicklaustrup committed:
Remove old type annotation syntax (Issue #27845)

This PR modernizes type annotations in the matrices/dense module.
Changes include:
- Added from __future__ import annotations
- Converted is_MatrixExpr = False # type: bool to is_MatrixExpr: bool = False

Tests pass, and code quality checks have been performed.

nicklaustrup committed:
Remove old type annotation syntax (Issue #27845)

This PR modernizes type annotations in the sympy/__init__.py module.
Changes include:
- Converting type comments to PEP 526 variable annotations
- Adding `from __future__ import annotations` for Python 3.8 compatibility

Tests pass, and code quality checks have been performed.

Signed-off-by: Nicholas Laustrup <124007393+nicklaustrup@users.noreply.github.com>

dbrnz committed:
mathml: add `disable_split_super_sub` option to disable name splitting

dbrnz committed:
Add self to `.mailmap`.

nicklaustrup committed:
Remove old type annotation syntax (Issue #27845)

This PR modernizes type annotations in the codegen/ast.py
and codegen/fnodes.py module.
Changes include:
- Converting type comments to PEP 526 variable annotations
- Adding `from __future__ import annotations` for Python 3.8 compatibility
- Removing redundant None initializations used only for type hinting

Tests pass, and code quality checks have been performed.

Signed-off-by: Nicholas Laustrup <124007393+nicklaustrup@users.noreply.github.com>

nicklaustrup committed:
Update old type annotation syntax (Issue #27845)

This PR modernizes type annotations in the core/evalf module.
Changes include:
- Converting type comments to PEP 526 variable annotations
- Replacing typing aliases (tDict, tTuple, etc.) with direct type references
- Adding `from __future__ import annotations` for Python 3.8 compatibility
- Removing redundant None initializations used only for type hinting

Tests pass, and code quality checks have been performed.

Signed-off-by: Nicholas Laustrup <124007393+nicklaustrup@users.noreply.github.com>

nicklaustrup committed:
Remove old type annotation syntax (Issue #27845)

This PR modernizes type annotations in the [module name] module.
Changes include:
- Replacing typing aliases (Dimension) with direct type references

Tests pass, and code quality checks have been performed.

Signed-off-by: Nicholas Laustrup <124007393+nicklaustrup@users.noreply.github.com>

nicklaustrup committed:
author: add Nicholas Laustrup to .mailmap

oscarbenjamin commented on sympy/core/evalf.py:
```diff
@@ -1235,7 +1234,7 @@ def check_convergence(numer: 'Expr', denom: 'Expr', n: 'Symbol') -> tTuple[int,
 
 
-def hypsum(expr: 'Expr', n: 'Symbol', start: int, prec: int) -> mpf:
+def hypsum(expr: 'Expr', n: 'Symbol', start: int, prec: int) -> mpf:  # type: ignore
```
Why is the type ignore needed here?

nicklaustrup replied:
The linter didn't like having the variable as a return, it was expecting a type instead. I can remove the type ignore if you think its not necessary

oscarbenjamin replied:
Whether it is necessary is just down to whether `mypy sympy` would otherwise reject the code. I'm assuming that mypy rejected the code and then I'm wondering what caused it to reject the code.

nicklaustrup replied:
Looks like mypy passes it without the ignore. I believe I added it because of a Pylance warning

oscarbenjamin commented:
> I'd like feedback on my approach before proceeding with other modules.

What you have here looks correct to me.

nicklaustrup committed:
typing: modernize annotations in matrices/common and matrixbase

Removed a "# type: ignore" from core/evalf.

Update type annotations in specified files following PEP 526:
- Convert "# type: X" comments to "variable: X" syntax.
- Replace typing aliases (tDict, tTuple, etc.) with built-in types.
- Add `from __future__ import annotations` for Python 3.8 compatibility.
- Remove redundant None initializations used only for typing.
- Convert Optional[X] to X | None and Union[A, B] to A | B.

Tests pass, and code quality checks have been performed.

Signed-off-by: Nicholas Laustrup <124007393+nicklaustrup@users.noreply.github.com>

dbrnz committed:
Tidy up: prefer conditional statement over conditional expression

nicklaustrup committed:
typing: modernize annotations in sympy/abc

Update type annotations in specified files following PEP 526:
- Replace typing aliases (tDict, tTuple, etc.) with built-in types.
- Add `from __future__ import annotations` for Python 3.8 compatibility.

Tests pass, and code quality checks have been performed.

Signed-off-by: Nicholas Laustrup <124007393+nicklaustrup@users.noreply.github.com>

nicklaustrup committed:
typing: modernize annotations in functions/elementary/trigonometric

Update type annotations in specified files following PEP 526:
- Convert "# type: X" comments to "variable: X" syntax.
- Replace typing aliases (tDict, tTuple, etc.) with built-in types.
- Add `from __future__ import annotations` for Python 3.8 compatibility.
- Remove redundant None initializations used only for typing.
- Convert Optional[X] to X | None and Union[A, B] to A | B.

Tests pass, and code quality checks have been performed.

Signed-off-by: Nicholas Laustrup <124007393+nicklaustrup@users.noreply.github.com>

nicklaustrup committed:
typing: modernize annotations in physics/units/unitsystem

Update type annotations in specified files following PEP 526:
- Convert "# type: X" comments to "variable: X" syntax.
- Replace typing aliases (tDict, tTuple, etc.) with built-in types.
- Add `from __future__ import annotations` for Python 3.8 compatibility.

Tests pass, and code quality checks have been performed.

Signed-off-by: Nicholas Laustrup <124007393+nicklaustrup@users.noreply.github.com>

nicklaustrup committed:
typing: modernize annotations in concrete/products

Update type annotations in specified files following PEP 526:
- Replace typing aliases (tDict, tTuple, etc.) with built-in types.
- Add `from __future__ import annotations` for Python 3.8 compatibility.

Tests pass, and code quality checks have been performed.

Signed-off-by: Nicholas Laustrup <124007393+nicklaustrup@users.noreply.github.com>

nicklaustrup committed:
typing: modernize annotations in /core /concrete

Update type annotations in specified files following PEP 526:
- Replace typing aliases (tDict, tTuple, etc.) with built-in types.
- Add `from __future__ import annotations` for Python 3.8 compatibility.

Tests pass, and code quality checks have been performed.

Signed-off-by: Nicholas Laustrup <124007393+nicklaustrup@users.noreply.github.com>

dbrnz committed:
latex: add `disable_split_super_sub` option to disable name splitting

nicklaustrup committed:
typing: modernize annotations in /external and /functions

Update type annotations in specified files following PEP 526:
- Replace typing aliases (tDict, tTuple, etc.) with built-in types.
- Add `from __future__ import annotations` for Python 3.8 compatibility.

Tests pass, and code quality checks have been performed.

Signed-off-by: Nicholas Laustrup <124007393+nicklaustrup@users.noreply.github.com>

nicklaustrup committed:
typing: modernize annotations in parsing/ polys/ stats/ tensor/

Modified:
- parsing/sympy_parser.py
- polys/matrices/domainmatrix.py
- stats/stochastic_process_types.py
- tensor/array/expressions/array_expressions.py
- tensor/array/expressions/from_array_to_matrix.py

tensor/array/expressions/from_array_to_matrix.py:
- In _find_trivial_matrices_rewrite:
	- Left three variables initialized to None to avoid
	UnboundLocalErrors.

Update type annotations in specified files following PEP 526:
- Convert "# type: X" comments to "variable: X" syntax.
- Replace typing aliases (tDict, tTuple, etc.) with built-in types.
- Add `from __future__ import annotations` for Python 3.8 compatibility.
- Remove redundant None initializations used only for typing.
- Convert Optional[X] to X | None and Union[A, B] to A | B.

Tests pass, and code quality checks have been performed.

Signed-off-by: Nicholas Laustrup <124007393+nicklaustrup@users.noreply.github.com>

nicklaustrup committed:
typing: modernize annotations in polys/ printing/ solvers/ stats/

Modified:
- polys/domains/gaussiandomains.py
- printing/printer.py
- solvers/diophantine/diophantine.py
- solvers/ode/single.py
- stats/rv.py
- tensor/tensor.py

Update type annotations in specified files following PEP 526:
- Convert "# type: X" comments to "variable: X" syntax.
- Replace typing aliases (tDict, tTuple, etc.) with built-in types.
- Add `from __future__ import annotations` for Python 3.8 compatibility.
- Remove redundant None initializations used only for typing.
- Convert Optional[X] to X | None and Union[A, B] to A | B.
- Added ' | None' to type for 'is_real: bool' in stats/rv.py::PSpace

Only remaining typing tType:
- from typing import Sequence as tSequence
- In sympy/stats/stochastic_process_types.py

Most tests pass, and code quality checks have been performed.
Failing local tests  in:
- printing/pretty/tests/test_pretty.py
- printing/tests/test_latex.py
- printing/tests/test_str.py
Due to:
- test_RandomDomain - AttributeError: 'SingleContinuousPSpace' object has no attribute 'is_Finite'
- test_latex_RandomDomain - AttributeError: 'SingleContinuousPSpace' object has no attribute 'is_Finite'
These tests were failing prior to making these changes.

Signed-off-by: Nicholas Laustrup <124007393+nicklaustrup@users.noreply.github.com>

nicklaustrup committed:
Initialized variables in stats/rv.py to fix failing tests

Tests will fail if these are not initialized to None:
**is_Finite**: bool | None = None
**is_Continuous**: bool | None = None
**is_Discrete**: bool | None = None

Tests are passing now.

Signed-off-by: Nicholas Laustrup <124007393+nicklaustrup@users.noreply.github.com>

oscarbenjamin commented on sympy/core/evalf.py:
```diff
@@ -1193,7 +1192,7 @@ def evalf_integral(expr: 'Integral', prec: int, options: OPT_DICT) -> TMP_RES:
 
 
-def check_convergence(numer: 'Expr', denom: 'Expr', n: 'Symbol') -> tTuple[int, Any, Any]:
+def check_convergence(numer: 'Expr', denom: 'Expr', n: 'Symbol') -> tuple[int, Any, Any]:
```
I think that we can also change `'Expr'` to `Expr` now because we have `__future__.annotations` and `if TYPE_CHECKING`.

dbrnz committed:
printing: extend `disable_split_super_sub` tests to include default and enabled cases.

oscarbenjamin committed:
lint: remove unneeded global statements

Commit:
Add caching for harmonic number computation

dbrnz committed:
Update entry in `.mailmap`.

nicklaustrup committed:
typing: modernize annotations in core/evalf

Replaced 'Expr' with Expr

Tests pass, and code quality checks have been performed.

Signed-off-by: Nicholas Laustrup <124007393+nicklaustrup@users.noreply.github.com>

nicklaustrup commented:
@oscarbenjamin I resolved a merge conflict but now I see my name on those commits... I am not sure if I messed something up, still learning the git workflow. Please advise if I need to fix something

nicklaustrup committed:
Merge branch 'master' into update-python-typehints

nicklaustrup committed:
Modified typehints in /__init__ and /core/evalf

- Removed TypeHint from SYMPY_DEBUG due to failing
python 2 test case.
- Added typing import tTuple back to MPF_TUP in
core/evalf.py due to failing python 3.8 test

Signed-off-by: Nicholas Laustrup <124007393+nicklaustrup@users.noreply.github.com>

nicklaustrup committed:
Initialized printmethod in /printing/printer.py

Initialized the printmethod variable printing/printer.py
in order to pass tests.

Signed-off-by: Nicholas Laustrup <124007393+nicklaustrup@users.noreply.github.com>

nicklaustrup committed:
Removed trailing whitespace in sympy/__init__

Signed-off-by: Nicholas Laustrup <124007393+nicklaustrup@users.noreply.github.com>

nicklaustrup committed:
Revert changes made to printing/printer

These changes were initially made to update
the file to modern Type annotations but the changes
seemed to not be compatible with some of the tests so they
are being reverted.

Signed-off-by: Nicholas Laustrup <124007393+nicklaustrup@users.noreply.github.com>

nicklaustrup committed:
Reverted changes to smypy/__init__ and core/evalf

These changes were failing the py2 and 3.8 tests respectively

Signed-off-by: Nicholas Laustrup <124007393+nicklaustrup@users.noreply.github.com>

nicklaustrup committed:
Reverting more changes to core/evalf

Signed-off-by: Nicholas Laustrup <124007393+nicklaustrup@users.noreply.github.com>

nicklaustrup committed:
Reverted more changes to core/evalf

Signed-off-by: Nicholas Laustrup <124007393+nicklaustrup@users.noreply.github.com>

nicklaustrup commented:
Sort of at an impasse here. Many of the type hinting and annotation changes are not compatible with the 3.8 test even with `from __future__ import annotations` so I'm unsure of how to proceed

oscarbenjamin commented:
> Many of the type hinting and annotation changes are not compatible with the 3.8 test even with `from __future__ import annotations` so I'm unsure of how to proceed

Can you give examples?

We don't need to support Python 3.8 so don't worry about it.

We should probably drop support for other Python versions as well but for now I'll just drop 3.8 from CI.

oscarbenjamin commented:
I've removed Python 3.8 from CI checks now.

nicklaustrup commented:
> I've removed Python 3.8 from CI checks now.

Okay in that case I can get these changes fixed up this evening

nicklaustrup committed:
Revert "Reverting more changes to core/evalf"

This reverts commit 5246711281aefef17677a61e1f9eded3ecd2119d.

nicklaustrup committed:
Revert "Reverted changes to smypy/__init__ and core/evalf"

This reverts commit 1b9dfe93833403166706b7dce44291529c5adcc5.

nicklaustrup committed:
Restored changes made that were failing 3.8 tests

Made changes in core/evalf and sympy/__init__.

Signed-off-by: Nicholas Laustrup <124007393+nicklaustrup@users.noreply.github.com>

nicklaustrup committed:
Removed tTuple from core/evalf.py

Signed-off-by: Nicholas Laustrup <124007393+nicklaustrup@users.noreply.github.com>

oscarbenjamin commented on sympy/core/evalf.py:
```diff
@@ -147,7 +147,7 @@ def fastlog(x: Optional[MPF_TUP]) -> tUnion[int, Any]:
 
 
-def pure_complex(v: 'Expr', or_real=False) -> tuple['Number', 'Number'] | None:
+def pure_complex(v: Expr, or_real=False) -> tuple['Number', 'Number'] | None:
```
I think that `'Number'` here can be `Number`.

oscarbenjamin commented on sympy/core/evalf.py:
```diff
@@ -1058,7 +1059,7 @@ def evalf_alg_num(a: 'AlgebraicNumber', prec: int, options: OPT_DICT) -> TMP_RES
 
 
-def as_mpmath(x: Any, prec: int, options: OPT_DICT) -> tUnion[mpc, mpf]:
+def as_mpmath(x: Any, prec: int, options: OPT_DICT) -> mpc | mpf:  # type: ignore
```
What is the `type: ignore` needed for?

nicklaustrup replied:
I added these two ignores because of a pylance error but I removed them and mypy passes.

oscarbenjamin commented on sympy/core/evalf.py:
```diff
@@ -1104,10 +1105,10 @@ def do_integral(expr: 'Integral', prec: int, options: OPT_DICT) -> TMP_RES:
+        max_imag_term: float | int = MINUS_INF
 
-        def f(t: 'Expr') -> tUnion[mpc, mpf]:
+        def f(t: Expr) -> mpc | mpf:  # type: ignore
```
What is `type: ignore` needed for?

oscarbenjamin commented on sympy/core/evalf.py:
```diff
@@ -1193,7 +1194,7 @@ def evalf_integral(expr: 'Integral', prec: int, options: OPT_DICT) -> TMP_RES:
 
 
-def check_convergence(numer: 'Expr', denom: 'Expr', n: 'Symbol') -> tTuple[int, Any, Any]:
+def check_convergence(numer: Expr, denom: Expr, n: 'Symbol') -> tuple[int, Any, Any]:
```
Can this be `Symbol`?

nicklaustrup replied:
Looks like it can be! Went ahead and updated them plus Numbers

oscarbenjamin commented on sympy/core/evalf.py:
```diff
@@ -1235,7 +1236,7 @@ def check_convergence(numer: 'Expr', denom: 'Expr', n: 'Symbol') -> tTuple[int,
 
 
-def hypsum(expr: 'Expr', n: 'Symbol', start: int, prec: int) -> mpf:
+def hypsum(expr: Expr, n: 'Symbol', start: int, prec: int) -> mpf:
```
Symbol?

oscarbenjamin commented on sympy/stats/rv.py:
```diff
@@ -196,10 +196,10 @@ class PSpace(Basic):
-    is_real = None  # type: bool
+    is_Finite: bool | None = None  # Fails test if not set to None
+    is_Continuous: bool | None = None  # Fails test if not set to None
+    is_Discrete: bool | None = None  # Fails test if not set to None
```
We do't have to do this here but I imagine this is just because of bogus test code that should be changed.

oscarbenjamin commented:
I left a few questions but otherwise this looks good.

nicklaustrup committed:
Updated type annotations in core/evalf

- Removed 2 extraneous # type: ignores
- Removed ' from 'Symbol' and 'Number'

Signed-off-by: Nicholas Laustrup <124007393+nicklaustrup@users.noreply.github.com>

nicklaustrup committed:
Updated type annotations for core/evalf

- Removed extra ' marks around Symbol and Number.
- Removed comment regarding failing test
- Removed unneeded type ignore comments

All tests passing.

Signed-off-by: Nicholas Laustrup <124007393+nicklaustrup@users.noreply.github.com>
